### PR TITLE
Validate publication and created years in work form

### LINKED ISSUE #238: Constrain and validate published & created years in work edit form
Add validation on the client-side backed with a server-side validation such that only four-digit years are allowable.

### PR DESCRIPTION
**Pull request by mjgiarlo**
Fixes #238

## Why was this change made?

To validate work date fields.

Includes:
* Add two new configuration options for the earliest publication and created years
* Display client-side validation errors in work form
* Back client-side validation with form-based validation
* Use reform type coercion and dry-types to ensure publication and created years are optional integers
* Create a custom type to allow the work form to handle blank strings, which is what the Rails form submits when an optional field is not filled it. (This is how reform uses dry-types to handle types.)
* Prefer Time.zone.today.year over Date.today.year (Rubocop said so)

## How was this change tested?

CI and in the browser.

## Which documentation and/or configurations were updated?

Added config (no shared_configs changes needed)

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -10,3 +10,14 @@
...
+# Offense count: 1
+# Cop supports --auto-correct.
+# Configuration parameters: SuggestedStrictness, Include.
+# Include: **/*.rb, **/*.rbi, **/*.rake, **/*.ru
+Sorbet/FalseSigil:
+  Exclude:
+    - 'bin/**/*'
+    - 'db/**/*.rb'
+    - 'script/**/*'
+    - 'config/initializers/dry_types.rb'
```
Comment (mjgiarlo) [inline .rubocop_todo.yml:23]: It didn't seem worth the time to work out how to get dry-types and sorbet to live together harmoniously.

File: app/components/works/dates_component.html.erb [hunk 2]
```diff
@@ -8,7 +8,8 @@
       <div class="row">
         <div class="col">
           <label for="work_published_year" class="visually-hidden">Publication year</label>
-          <%= number_field_tag 'work[published(1i)]', nil, id: 'work_published_year', class: "form-control", max: Date.today.year %>
+          <%= number_field_tag 'work[published(1i)]', nil, id: 'work_published_year', class: "form-control", min: Settings.earliest_publication_year, max: Time.zone.today.year %>
```
Comment (mjgiarlo) [inline app/components/works/dates_component.html.erb:11]: This was a super easy validation. Just needed to add a `min:` param to `number_field_tag`. Very cool. Haven't seen this sort of built-in browser/HTML validation before.

Then it was just a matter of include the `invalid-feedback` `div` immediately below with a corresponding error message. Super easy client-side validation without any fuss.

File: app/forms/work_form.rb [hunk 3]
```diff
@@ -19,18 +19,18 @@ class WorkForm < Reform::Form
...
-  property 'created_range(1i)', virtual: true
-  property 'created_range(2i)', virtual: true
-  property 'created_range(3i)', virtual: true
-  property 'created_range(4i)', virtual: true
-  property 'created_range(5i)', virtual: true
-  property 'created_range(6i)', virtual: true
-  property 'published(1i)', virtual: true
-  property 'published(2i)', virtual: true
-  property 'published(3i)', virtual: true
+  property 'created(1i)', virtual: true, type: ::Types::Custom::NilableInteger
```
Comment (mjgiarlo) [inline app/forms/work_form.rb:22]: See comment in `config/initializers/dry_types.rb` as to what this type does and why I defined it.

File: app/forms/work_form.rb [hunk 4]
```diff
@@ -47,6 +47,12 @@ def sync(*)
   end
 
   validates :title, presence: true
+  validates 'created(1i)', 'created_range(1i)', 'created_range(4i)',
+            inclusion: { in: Settings.earliest_created_year..Time.zone.today.year },
```
Comment (mjgiarlo) [inline app/forms/work_form.rb:51]: Back client-side validation with a form-based validation.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**